acpi-tables.bb extended with package generation and scripts

htot opened the pull request:
@andy-shev This is on top of the patches I added earlier to include edison sample tables.

Heads up - I rebased today on westeri/meta-acpi/master.

I extended the original recipe with building a package (the package type is selected elsewhere in the config, YP default is ipk but we use deb). This does not change anything to the generated cpio. 

The generated package sits harmless in the YP build system and is only installed when added to the system image edison-image.bb using IMAGE_INSTALL += "acpi-tables" or better IMAGE_INSTALL_append as I learned from @alext-mkrs.

This also installs a systemd service that can be enabled to start on boot and executes a script that copies all acpi-tables to configfs.

Additional aml's can also be copied manually to /kernel/firmware/acpi/ and will be loaded into the kernel on the next reboot. Also I think users should be able to easily bbappend the recipy to add their own asl tables for inclusion into their system image.

I tested this by manually building `'bitbake acpi-tables'`, copying the `acpi-tables_1.0-r0_amd64.deb` over to edison and install using `'dpkg -i acpi-tables_1.0-r0_amd64.deb'`

With this extension to the meta-acpi layer and configfs enabled in the kernel we have acpi support working for edison. Functionally nothing else is needed. I hope @alext-mkrs is willing to review this.

Also this same method should work on the other supported platforms as well when configfs is enabled, it is independent of the boot loader and kernel image generation.

htot committed:
edison: Add acpi sample tables

Tables authored by Andy Shevchenko<andriy.shevchenko@linux.intel.com>
Retrieved from https://edison.internet-share.com/wiki/ACPI

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

htot committed:
acpi-tables: build package

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

htot committed:
acpi-tables: add script and systemd service

This installs a script acpi-tables-load that can be manually run to load all acpi tables in /kernel/firmware/acpi
on a configfs enabled kernel. The script assumes configfs is mounted on /sys/kernel/config as is the default.
Further a systemd service acpi-tables-load.service is installed that runs the script . To run once manually,
use 'systemctl start acpi-tables-load', to automatically run after boot 'systemctl enable acpi-tables-load'

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>

alext-mkrs commented:
Sure, I'll take a look. At the first glance - systemd stuff (1) doesn't IMHO fully fit the meta-acpi philosophy and (2) should be implemented using "systemd" Yocto class, which checks whether systemd is enabled at all, provides standard facilities for installing unit files, etc.

htot commented:
Ok. (1) I don't think I fully understand the meta-acpi intention, probably because I don't have any of the currently supported boards. But if the philosophy is adding acpi to an existing bsp, that is what I attempted here.
(2) I just followed existing examples from meta-intel-edison, maybe those are outdated. I'll catch up on YP's systemd concept.

htot commented:
@alext-mkrs on (1) do I understand this correctly: meta-intel acts as the bsp and uses rmc to create a configurable bsp for multiple boards? If so, we should be working to mould edison u-boot, kernel and image building to use that? Best place to add these things for now would be meta-intel-edison until we get that stable, and then see if we can get it into meta-intel or meta-intel-contrib. Imho meta-acpi would not be the best place to add f.i. kernels. In fact meta-intel already has 4.14 (in rocko). meta-acpi can be (in time) be reduced to only adding acpi tables late (either from cpio or through configfs). Eventually we should also be able to reduce meta-intel-edison to containing just a few edison specific recipies.  The patches above could be considered a first step in this direction.

Problem with rmc is, I have no idea how that works. Maybe you can take the lead in this area and I will follow? Maybe after fixing up i2c?

On (2) I agree that the above patches can be improved to utilize the systemd class. Otoh they are not invalid as they are. With your help and a good example (have been looking at connman.bb) I should be able to fix this up. So then, pull into meta-acpi now and improve later? Or leave in my repo until improved.

andy-shev commented:
The kernel in meta-acpi is built b/c at the time it was started meta-intel was quite behind the mainline.
So, I'm fine to accept a patch that removes kernel build from meta-acpi (or rather makes it optional and disabled for time being).

U-Boot on the other hand is quite specific to Edison (and some of the boards which it actually supports, like Minnowboard MAX).

So, at the end (ideal case) U-Boot should be also part of meta-intel. Though we might split it to a separate layer for now (meta-intel-u-boot ?) or do a clean machine dependent build inside meta-acpi.

What I would rather like to see first is to establishment of *edison* machine here. First step just basics, no special requirements or dependencies. Meanwhile we need to have U-Boot, and add this dependency to *edison* machine later on (as a second step).

This PR is no go, sorry.

alext-mkrs commented:
Gentlemen, this week turned out to be more hectic than expected, so I've got no spare time at all. I'll go over your comments over the weekend and get back with my comments.

htot commented:
@andy-shev edison machine is already defined and needed (but in meta-intel-edison). It defines variables used by linux-yocto and u-boot, I think that is not a problem, as those are destined to be needed. But the only versions of these recipes implementing a meaningful result (and replacing the one in poky) are in meta-intel-edison. 

The edison.conf also defines variables for INITRAMFS generation. The poky recipe for initramfs is appended by meta-intel-edison.

Other than the above I see no problem in moving edison.conf to meta-acpi (which will make meta-acpi a dependency for meta-intel-edison also in the non-acpi case).

But maybe it's possible to have a basic definition on edison in meta-intel and extend that in meta-intel-edison.

I think @alext-mkrs has a much more informed opinion on this than I have.

Before we consider moving this we would need to clean the lines with FIXME (fixme from original) and lines related g_multi (I don't know what module g_multi does and probably disabled this in a very early stage).

And the systemd recipe should be fortified as @alext-mkrs remarked.

alext-mkrs commented:
Okay, so let me try to summarize our discussions here and in https://github.com/alext-mkrs/meta-acpi/pull/1 and propose the next steps.

1. `meta-acpi` is rather *not* envisioned as a BSP-type layer. It ideally shouldn't be building kernel or U-Boot, or anything similar, directly. It should depend on respective BSP layers like `meta-intel` for that. It should only focus on adding ACPI functionality and contain only ACPI-related full recipes/overrides/appends. Those should be written so as to detect which `MACHINE` we build for right now, and - if it's supported by `meta-acpi` - build respective ACPI tables and whatever else is needed to enable ACPI on that specific machine;
    1. To that end, given that `meta-intel`'s kernel is new enough these days, we want to remove kernel-building recipes from `meta-acpi` (disable by default - as a first step);
    1. There's an open question of what to do with U-Boot, as there's no support for it in `meta-intel` right now and builds may be quite board-specific. However if the main idea described above is implemented, for `meta-acpi` it will be "don't care" - it should just provide necessary overrides/appends to build an ACPI-enabled U-Boot _if_ it is enabled in the BSP layer in use (the list of supported ones should be explicit);
1. We want to add Edison machine support to the layer, along with the respective ACPI table samples;

So my proposal is to create a set of records in this repo's issues tracker for:
1. Kernel build removal (making it optional and switched off by default);
1. Adding Edison `MACHINE` support - only sample ACPI tables and a way to bundle/activate them. Maybe in a way @htot proposed, loading through configfs + systemd service to do that. Alternative is loading from initramfs, but it's not clear whether that's possible (initial tests were unsuccessful). If systemd way is used - the recipe must use Yocto's `systemd` class to do proper build-time feature detection and unit file deployment without doing all of that manually;
1. Adding U-Boot overrides/appends for enabling ACPI build (and removing this piece from `meta-intel-edison`), just for Edison as the first step;
1. Adding kernel overrides/appends for enabling ACPI build (and removing this piece from `meta-intel-edison`), just for Edison as the first step;
1. Adding kernel overrides/appends for enabling ACPI build (and removing this piece from `meta-intel-edison`) for the rest of the boards supported by `meta-acpi` (longer-term one);

Does that make sense? What do you think? If we are in a consensus, I'll create those trackers and we'll be able to discuss specific technical solutions for those there, picking stuff from this PR and https://github.com/alext-mkrs/meta-acpi/pull/1.

htot commented:
On 2: I just discovered it is not only possible to load acpi table, but also unload them through configfs. That opens the possibility to have loading/unloading done by mraa on demand, keeping compatibility with existing mraa application. For this we need separate acpi tables for each device we want to enable (spi, led, uart).
Also, I tested loading from initramfs does not work, the kernel documentation is correct, although this might change in the future. They need to be loaded from an uncompressed cpio (so external from the initramfs), loaded separately by u-boot. This cpio is already built by meta-acpi. Loading via this method works fine.
IMHO loading the cpio is probably best for a finished application (let's say an IOT gateway based on edison), while through configfs (and possibly mraa) for a development / maker setup.
1 we can split out meta-intel-edison-bsp from meta-intel-edison and migrate in the direction for inclusion or as an extension to meta-intel.

andy-shev commented:
@alext-mkrs

1. You got it completely right! U-Boot, unfortunately, needs to be build either way in the first place. So, I would like to see in `meta-intel` rather U-Boot/ACPI than just U-Boot. The rationale is that all Intel IoT platforms are ACPI enabled with exception of Edison. So, it's natural to assume that all of them ACPI from `meta-intel` prospective. But in any case we can start from some simple steps.

2. Correct!

With regard to road map:
1. Yes.
2. `meta-acpi` is designed to support only initramfs approach for now. So, I'm totally to support configfs approach, but let's do this generic (good start, even before Edison) is to enable it for Galileo Gen 2.
3. As a first steps (see above).
4. Yes, `meta-intel-edison`, from my point of view, should be clean out of ACPI related things and be just a continuation of not-so-standard-intel-platform-bsp. Of course, if you decide otherwise I'm not going to stop you.
5. Yes.

htot commented:
2. Package we generate right now should already work as long as the kernel enabled configfs support (that needs to be added) and nothing in SSDT is needed for booting. Needs testing of course.
4. We already removed lot's of stuff from meta-intel-edison, as they were either moved upstream, or upstream updated to new versions. Or I just removed because didn't work or had become obsolete. Once we get kernel and u-boot out, and a few other things, it will just be one example (reference?) image for edison. Probably the same image could run on galileo just by selecting the machine. That would be a good thing. The addition of acpi right now consists only of flipping a switch, which selects a different commit of u-boot and kernel. The switch is in:
`DISTRO_FEATURES = "systemd alsa argp bluetooth ext2 largefile usbgadget usbhost wifi xattr nfs zeroconf pci pam ${DISTRO_FEATURES_LIBC}"`.
Adding acpi simply activates u-boot and acpi enabled kernel. By default non-acpi version are built.
